pkg/provisioner/ironic/bios.go

```diff
@@ -0,0 +1,171 @@
+package ironic
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+)
+
+// BuildBIOSCleanSteps build clean steps for different driver
+func BuildBIOSCleanSteps(driver string, firmware *metal3v1alpha1.FirmwareConfig) (cleanSteps []nodes.CleanStep, err error) {
+	// Deal possible panic
+	defer func() {
+		r := recover()
+		if r != nil {
+			err = fmt.Errorf("panic in build BIOS settings: %v", r)
+		}
+	}()
+
+	switch driver {
+	case "ilo", "ilo5":
+		cleanSteps = buildILOCleanSteps(firmware)
+	case "irmc":
+		cleanSteps = buildIRMCCleanSteps(firmware)
+	default:
+		if firmware != nil {
+			return nil, fmt.Errorf("Firmware settings are defined, but the node's driver %s does not support configure firmware setting", driver)
```

Comment:
Let's reword this to "firmware settings for driver %s are not supported". Otherwise you need to add support for Redfish in the same PR (Redfish has BIOS settings).

Reply:
Does BIOS settings depend on BMC when using the redfish API? If yes, where can I get the BMC‘s type? Or we can define the redfish BIOS settings by following yaml:
```yaml
firmware:
  redfish:
     bmcType: "irmc" # Allow value: idrac, irmc, ilo  required
     ...... # BIOS settings
```